// @nrwl/workspace/src/tasks-runner/run-command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRunner = exports.createTask = exports.createTasksForProjectToRun = exports.runCommand = void 0;
const tslib_1 = require("tslib");
const path_1 = require("path");
const app_root_1 = require("@nrwl/tao/src/utils/app-root");
const devkit_1 = require("@nrwl/devkit");
const logger_1 = require("@nrwl/tao/src/shared/logger");
const fileutils_1 = require("../utilities/fileutils");
const project_graph_utils_1 = require("../utilities/project-graph-utils");
const output_1 = require("../utilities/output");
const utils_1 = require("./utils");
const life_cycle_1 = require("./life-cycle");
const run_many_terminal_output_life_cycle_1 = require("./run-many-terminal-output-life-cycle");
const empty_terminal_output_life_cycle_1 = require("./empty-terminal-output-life-cycle");
const run_one_terminal_output_life_cycle_1 = require("./run-one-terminal-output-life-cycle");
const task_timings_life_cycle_1 = require("./task-timings-life-cycle");
function getTerminalOutputLifeCycle(initiatingProject, terminalOutputStrategy, projectsToRun, tasks, nxArgs, overrides) {
    if (terminalOutputStrategy === 'run-one') {
        return new run_one_terminal_output_life_cycle_1.RunOneTerminalOutputLifeCycle(initiatingProject, projectsToRun.map((t) => t.name), tasks, nxArgs);
    }
    else if (terminalOutputStrategy === 'hide-cached-output') {
        return new empty_terminal_output_life_cycle_1.EmptyTerminalOutputLifeCycle();
    }
    else {
        return new run_many_terminal_output_life_cycle_1.RunManyTerminalOutputLifeCycle(projectsToRun.map((t) => t.name), tasks, nxArgs, overrides);
    }
}
function runCommand(projectsToRun, projectGraph, { nxJson, workspaceResults }, nxArgs, overrides, terminalOutputStrategy, initiatingProject) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
        if (nxArgs.onlyFailed || nxArgs['only-failed']) {
            output_1.output.warn({
                title: '--only-failed has been removed. Nx is executing all the tasks.',
            });
        }
        const { tasksRunner, runnerOptions } = getRunner(nxArgs, nxJson);
        // Doing this for backwards compatibility, should be removed in v14
        ensureTargetDependenciesBackwardCompatibility(nxJson, nxArgs);
        const defaultDependencyConfigs = nxJson.targetDependencies;
        const tasks = createTasksForProjectToRun(projectsToRun, {
            target: nxArgs.target,
            configuration: nxArgs.configuration,
            overrides,
        }, projectGraph, initiatingProject, defaultDependencyConfigs);
        const lifeCycles = [
            getTerminalOutputLifeCycle(initiatingProject, terminalOutputStrategy, projectsToRun, tasks, nxArgs, overrides),
        ];
        if (process.env.NX_PERF_LOGGING) {
            lifeCycles.push(new task_timings_life_cycle_1.TaskTimingsLifeCycle(process.env.NX_PERF_LOGGING));
        }
        const lifeCycle = new life_cycle_1.CompositeLifeCycle(lifeCycles);
        const promiseOrObservable = tasksRunner(tasks, Object.assign(Object.assign({}, runnerOptions), { lifeCycle }), {
            initiatingProject,
            target: nxArgs.target,
            projectGraph,
            nxJson,
        });
        let anyFailures;
        if (promiseOrObservable.subscribe) {
            anyFailures = yield anyFailuresInObservable(promiseOrObservable);
        }
        else {
            // simply await the promise
            anyFailures = yield anyFailuresInPromise(promiseOrObservable);
        }
        // fix for https://github.com/nrwl/nx/issues/1666
        if (process.stdin['unref'])
            process.stdin.unref();
        process.exit(anyFailures ? 1 : 0);
    });
}
exports.runCommand = runCommand;
function anyFailuresInPromise(promise) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
        return Object.values(yield promise).some((v) => v === 'failure' || v === 'skipped');
    });
}
function anyFailuresInObservable(obs) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
        return yield new Promise((res) => {
            let anyFailures = false;
            obs.subscribe((t) => {
                if (!t.success) {
                    anyFailures = true;
                }
            }, (error) => {
                output_1.output.error({
                    title: 'Unhandled error in task executor',
                });
                console.error(error);
                res(true);
            }, () => {
                res(anyFailures);
            });
        });
    });
}
function createTasksForProjectToRun(projectsToRun, params, projectGraph, initiatingProject, defaultDependencyConfigs = {}) {
    var _a, _b;
    const tasksMap = new Map();
    const seenSet = new Set();
    for (const project of projectsToRun) {
        addTasksForProjectTarget(Object.assign(Object.assign({ project }, params), { errorIfCannotFindConfiguration: project.name === initiatingProject }), defaultDependencyConfigs, projectGraph, (_b = (_a = project.data.targets) === null || _a === void 0 ? void 0 : _a[params.target]) === null || _b === void 0 ? void 0 : _b.executor, tasksMap, [], seenSet);
    }
    return Array.from(tasksMap.values());
}
exports.createTasksForProjectToRun = createTasksForProjectToRun;
function addTasksForProjectTarget({ project, target, configuration, overrides, errorIfCannotFindConfiguration, }, defaultDependencyConfigs = {}, projectGraph, originalTargetExecutor, tasksMap, path, seenSet) {
    var _a, _b;
    const task = createTask({
        project,
        target,
        configuration,
        overrides: ((_b = (_a = project.data.targets) === null || _a === void 0 ? void 0 : _a[target]) === null || _b === void 0 ? void 0 : _b.executor) === originalTargetExecutor
            ? overrides
            : {},
        errorIfCannotFindConfiguration,
    });
    const dependencyConfigs = (0, utils_1.getDependencyConfigs)({ project: project.name, target }, defaultDependencyConfigs, projectGraph);
    if (dependencyConfigs) {
        for (const dependencyConfig of dependencyConfigs) {
            addTasksForProjectDependencyConfig(project, {
                target,
                configuration,
                overrides,
            }, dependencyConfig, defaultDependencyConfigs, projectGraph, originalTargetExecutor, tasksMap, path, seenSet);
        }
    }
    tasksMap.set(task.id, task);
}
function createTask({ project, target, configuration, overrides, errorIfCannotFindConfiguration, }) {
    var _a, _b;
    if (!(0, project_graph_utils_1.projectHasTarget)(project, target)) {
        output_1.output.error({
            title: `Cannot find target '${target}' for project '${project.name}'`,
        });
        process.exit(1);
    }
    configuration !== null && configuration !== void 0 ? configuration : (configuration = (_b = (_a = project.data.targets) === null || _a === void 0 ? void 0 : _a[target]) === null || _b === void 0 ? void 0 : _b.defaultConfiguration);
    const config = (0, project_graph_utils_1.projectHasTargetAndConfiguration)(project, target, configuration)
        ? configuration
        : undefined;
    if (errorIfCannotFindConfiguration && configuration && !config) {
        output_1.output.error({
            title: `Cannot find configuration '${configuration}' for project '${project.name}:${target}'`,
        });
        process.exit(1);
    }
    const qualifiedTarget = {
        project: project.name,
        target,
        configuration: config,
    };
    return {
        id: getId(qualifiedTarget),
        target: qualifiedTarget,
        projectRoot: project.data.root,
        overrides: interpolateOverrides(overrides, project.name, project.data),
    };
}
exports.createTask = createTask;
function addTasksForProjectDependencyConfig(project, { target, configuration, overrides, }, dependencyConfig, defaultDependencyConfigs, projectGraph, originalTargetExecutor, tasksMap, path, seenSet) {
    const targetIdentifier = getId({
        project: project.name,
        target,
        configuration,
    });
    seenSet.add(project.name);
    if (path.includes(targetIdentifier)) {
        output_1.output.error({
            title: `Could not execute ${path[0]} because it has a circular dependency`,
            bodyLines: [`${[...path, targetIdentifier].join(' --> ')}`],
        });
        process.exit(1);
    }
    if (tasksMap.has(targetIdentifier)) {
        return;
    }
    if (dependencyConfig.projects === 'dependencies') {
        const dependencies = projectGraph.dependencies[project.name];
        if (dependencies) {
            for (const dep of dependencies) {
                const depProject = projectGraph.nodes[dep.target] ||
                    projectGraph.externalNodes[dep.target];
                if ((0, project_graph_utils_1.projectHasTarget)(depProject, dependencyConfig.target)) {
                    addTasksForProjectTarget({
                        project: depProject,
                        target: dependencyConfig.target,
                        configuration,
                        overrides,
                        errorIfCannotFindConfiguration: false,
                    }, defaultDependencyConfigs, projectGraph, originalTargetExecutor, tasksMap, [...path, targetIdentifier], seenSet);
                }
                else {
                    if (seenSet.has(dep.target)) {
                        continue;
                    }
                    addTasksForProjectDependencyConfig(depProject, { target, configuration, overrides }, dependencyConfig, defaultDependencyConfigs, projectGraph, originalTargetExecutor, tasksMap, path, seenSet);
                }
            }
        }
    }
    else {
        addTasksForProjectTarget({
            project,
            target: dependencyConfig.target,
            configuration,
            overrides,
            errorIfCannotFindConfiguration: true,
        }, defaultDependencyConfigs, projectGraph, originalTargetExecutor, tasksMap, [...path, targetIdentifier], seenSet);
    }
}
function getId({ project, target, configuration, }) {
    let id = `${project}:${target}`;
    if (configuration) {
        id += `:${configuration}`;
    }
    return id;
}
function getRunner(nxArgs, nxJson) {
    let runner = nxArgs.runner;
    //TODO: vsavkin remove in Nx 12
    if (!nxJson.tasksRunnerOptions) {
        const t = require('./default-tasks-runner');
        return {
            tasksRunner: t.defaultTasksRunner,
            runnerOptions: nxArgs,
        };
    }
    //TODO: vsavkin remove in Nx 12
    if (!runner && !nxJson.tasksRunnerOptions.default) {
        const t = require('./default-tasks-runner');
        return {
            tasksRunner: t.defaultTasksRunner,
            runnerOptions: nxArgs,
        };
    }
    runner = runner || 'default';
    if (nxJson.tasksRunnerOptions[runner]) {
        let modulePath = nxJson.tasksRunnerOptions[runner].runner;
        let tasksRunner;
        if (modulePath) {
            if ((0, fileutils_1.isRelativePath)(modulePath)) {
                modulePath = (0, path_1.join)(app_root_1.appRootPath, modulePath);
            }
            tasksRunner = require(modulePath);
            // to support both babel and ts formats
            if (tasksRunner.default) {
                tasksRunner = tasksRunner.default;
            }
        }
        else {
            tasksRunner = require('./default-tasks-runner').defaultTasksRunner;
        }
        return {
            tasksRunner,
            runnerOptions: Object.assign(Object.assign({}, nxJson.tasksRunnerOptions[runner].options), nxArgs),
        };
    }
    else {
        output_1.output.error({
            title: `Could not find runner configuration for ${runner}`,
        });
        process.exit(1);
    }
}
exports.getRunner = getRunner;
function interpolateOverrides(args, projectName, projectMetadata) {
    const interpolatedArgs = Object.assign({}, args);
    Object.entries(interpolatedArgs).forEach(([name, value]) => {
        if (typeof value === 'string') {
            const regex = /{project\.([^}]+)}/g;
            interpolatedArgs[name] = value.replace(regex, (_, group) => {
                if (group.includes('.')) {
                    throw new Error('Only top-level properties can be interpolated');
                }
                if (group === 'name') {
                    return projectName;
                }
                return projectMetadata[group];
            });
        }
    });
    return interpolatedArgs;
}
function ensureTargetDependenciesBackwardCompatibility(nxJson, nxArgs) {
    var _a;
    (_a = nxJson.targetDependencies) !== null && _a !== void 0 ? _a : (nxJson.targetDependencies = {});
    if (nxArgs.withDeps) {
        devkit_1.logger.warn((0, logger_1.stripIndent)(`
        DEPRECATION WARNING: --with-deps is deprecated and it will be removed in v14.
        Configure target dependencies instead: https://nx.dev/latest/angular/core-concepts/configuration#target-dependencies.
      `));
        if (!nxJson.targetDependencies[nxArgs.target]) {
            nxJson.targetDependencies[nxArgs.target] = [
                { target: nxArgs.target, projects: 'dependencies' },
            ];
        }
    }
}
