// @nrwl/workspace/src/tasks-runner/task-timings-life-cycle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskTimingsLifeCycle = void 0;
const fs_1 = require("fs");
const perf_hooks_1 = require("perf_hooks");
const path_1 = require("path");
class TaskTimingsLifeCycle {
    constructor(perfLoggingEnvValue) {
        this.perfLoggingEnvValue = perfLoggingEnvValue;
        this.timings = {};
        this.profile = [];
        this.registeredGroups = new Set();
        if (typeof perfLoggingEnvValue === 'string' &&
            perfLoggingEnvValue !== 'true') {
            this.profileFile = (0, path_1.join)(process.cwd(), perfLoggingEnvValue);
        }
    }
    startTasks(tasks, { groupId }) {
        if (this.profileFile && !this.registeredGroups.has(groupId)) {
            this.registerGroup(groupId);
        }
        for (let t of tasks) {
            this.timings[`${t.target.project}:${t.target.target}`] = {
                start: new Date().getTime(),
                end: undefined,
                perfStart: perf_hooks_1.performance.now(),
            };
        }
    }
    endTasks(taskResults, metadata) {
        for (let tr of taskResults) {
            this.timings[`${tr.task.target.project}:${tr.task.target.target}`].end =
                new Date().getTime();
            this.timings[`${tr.task.target.project}:${tr.task.target.target}`].perfEnd = perf_hooks_1.performance.now();
        }
        if (this.profileFile) {
            this.recordTaskCompletions(taskResults, metadata);
        }
    }
    endCommand() {
        console.log('Task Execution Timings:');
        const timings = {};
        Object.keys(this.timings).forEach((p) => {
            const t = this.timings[p];
            timings[p] = t.end ? t.end - t.start : null;
        });
        console.log(JSON.stringify(timings, null, 2));
        if (this.profileFile) {
            (0, fs_1.writeFileSync)('profile.json', JSON.stringify(this.profile));
            console.log(`Performance Profile: ${this.profileFile}`);
        }
    }
    recordTaskCompletions(tasks, { groupId }) {
        for (const { task, status } of tasks) {
            const { perfStart, perfEnd } = this.timings[`${task.target.project}:${task.target.target}`];
            this.profile.push({
                name: task.id,
                cat: Object.values(task.target).join(','),
                ph: 'X',
                ts: perfStart * 1000,
                dur: (perfEnd - perfStart) * 1000,
                pid: process.pid,
                tid: groupId,
                args: {
                    target: task.target,
                    status,
                },
            });
        }
    }
    registerGroup(groupId) {
        this.profile.push({
            name: 'thread_name',
            ph: 'M',
            pid: process.pid,
            tid: groupId,
            ts: 0,
            args: {
                name: 'Group #' + (groupId + 1),
            },
        });
        this.registeredGroups.add(groupId);
    }
}
exports.TaskTimingsLifeCycle = TaskTimingsLifeCycle;
